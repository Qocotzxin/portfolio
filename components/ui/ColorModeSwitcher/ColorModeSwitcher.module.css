.stars {
    animation: rotation 10s infinite linear;
}

.cloud {
    animation: float 3s infinite linear;
    animation-direction: alternate;
}

@keyframes rotation {
    to {
        transform: rotate(360deg);
    }
}

@keyframes float {
    from {
        transform: translateX(0);
    }

    to {
        transform: translateX(8px);
    }
}