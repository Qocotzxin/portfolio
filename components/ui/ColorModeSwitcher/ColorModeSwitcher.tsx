import { Box, Icon, useColorMode } from "@chakra-ui/react";
import { useHygraphContext } from "@context/hygraphContext";
import { FC, useEffect, useState } from "react";
import { BsFillCloudsFill, BsStars } from "react-icons/bs";
import styles from "./ColorModeSwitcher.module.css";

const ColorModeSwitcher: FC = () => {
  const { ariaLabels } = useHygraphContext();
  const { colorMode, toggleColorMode } = useColorMode();
  const [isDark, setIsDark] = useState(colorMode === "dark");

  useEffect(() => {
    setIsDark(colorMode === "dark");
  }, [colorMode]);

  const onToggle = () => {
    setIsDark((isDark) => !isDark);
    setTimeout(toggleColorMode, 300);
  };

  return (
    <Box
      order="100"
      as="button"
      w="64px"
      h="32px"
      bg={isDark ? "blue.700" : "blue.200"}
      borderRadius="18px"
      _dark={{ borderColor: "gray.50" }}
      onClick={onToggle}
      position="relative"
      aria-label={ariaLabels?.find((al) => al.metadata !== colorMode)?.content}
    >
      {/* Visually hidden circle */}
      <Box
        borderRadius="50%"
        bg={isDark ? "blue.700" : "blue.200"}
        w="18px"
        h="18px"
        position="absolute"
        left="0"
        top="7px"
        zIndex="3"
      />
      {/* Main circle (sun/moon) */}
      <Box
        borderRadius="50%"
        bg="yellow.400"
        w="18px"
        h="18px"
        position="absolute"
        left="6px"
        top="7px"
        transform={`translateX(${isDark ? 0 : 34}px)`}
        transition="transform 0.5s"
        zIndex="2"
      />
      {/* Stars */}
      <Icon
        h="16px"
        w="16px"
        as={BsStars}
        position="absolute"
        right="8px"
        top="7px"
        color="white"
        opacity={isDark ? 1 : 0}
        transition="opacity 0.3s"
        className={styles.stars}
        zIndex="1"
      />
      {/* Clouds */}
      <Icon
        h="16px"
        w="16px"
        position="absolute"
        left="8px"
        top="7px"
        color="white"
        as={BsFillCloudsFill}
        opacity={isDark ? 0 : 1}
        transition="opacity 0.4s"
        zIndex="4"
        className={styles.cloud}
      />
    </Box>
  );
};

export default ColorModeSwitcher;
