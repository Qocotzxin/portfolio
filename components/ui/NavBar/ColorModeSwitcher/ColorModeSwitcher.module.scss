.ColorModeSwitcher {
  width: 64px;
  height: 32px;
  border-radius: 18px;
  position: relative;
  background: var(--ce-color-blue-200);

  &--dark {
    background: var(--ce-color-blue-700);
    border-color: var(--ce-color-gray-50);
  }

  &__circle {
    border-radius: 50%;
    width: 18px;
    height: 18px;
    position: absolute;

    &-outter {
      background: var(--ce-color-blue-200);
      left: 0;
      top: 7px;
      z-index: 3;
      &--dark {
        background: var(--ce-color-blue-700);
      }
    }

    &-inner {
      background: var(--ce-color-yellow-400);
      left: 6px;
      top: 7px;
      z-index: 2;
      transform: translateX(34px);
      transition: transform 0.5s;

      &--dark {
        transform: translateX(0);
      }
    }
  }

  &__icon {
    height: 16px;
    width: 16px;
    position: absolute;
    right: 8px;
    top: 7px;
    color: var(--ce-color-white);

    &-stars {
      animation: rotation 15s infinite linear;
      opacity: 0;
      transition: opacity 0.3s;
      z-index: 1;

      &--dark {
        opacity: 1;
      }
    }

    &-clouds {
      fill: var(--ce-color-white);
      opacity: 1;
      transition: opacity 0.4s;
      z-index: 4;
      animation: float 3s infinite linear;
      animation-direction: alternate;
      left: 8px;
      right: unset;

      &--dark {
        opacity: 0;
      }
    }
  }
}

@keyframes rotation {
  to {
    transform: rotate(360deg);
  }
}

@keyframes float {
  from {
    transform: translateX(0);
  }

  to {
    transform: translateX(8px);
  }
}
