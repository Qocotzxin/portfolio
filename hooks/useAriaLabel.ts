import { useHygraphContext } from "@contexts/hygraphContext";
import { AriaLabel } from "@models/hygraph";
import filter from "lodash/filter";

/**
 *  Returns an AriaLabel when it finds a coincidence based on the conditions provided. Else it returns null.
 */
export function useAriaLabel(conditions: Partial<AriaLabel>): AriaLabel | null {
  const { ariaLabels } = useHygraphContext();
  const results = filter(ariaLabels, conditions);

  return results.length ? results[0] : null;
}
