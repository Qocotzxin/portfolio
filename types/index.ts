export enum ColorMode {
  light = "light",
  dark = "dark",
}
