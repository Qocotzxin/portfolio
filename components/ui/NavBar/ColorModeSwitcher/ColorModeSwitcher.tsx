import { Box, Icon, useColorMode } from "@chakra-ui/react";
import { useAriaLabel } from "@hooks/useAriaLabel";
import { Clouds, Stars } from "@icons";
import { Components } from "@models/hygraph";
import { FC, useEffect, useState } from "react";
import styles from "./ColorModeSwitcher.module.css";

const ColorModeSwitcher: FC = () => {
  const { colorMode, toggleColorMode } = useColorMode();
  const ariaLabel = useAriaLabel({
    component: Components.ColorModeSwitcher,
    metadata: colorMode === "dark" ? "light" : "dark",
  });
  const [isDark, setIsDark] = useState(colorMode === "dark");

  useEffect(() => {
    setIsDark(colorMode === "dark");
  }, [colorMode]);

  const onToggle = () => {
    setIsDark((isDark) => !isDark);
    setTimeout(toggleColorMode, 300);
  };

  return (
    <Box
      as="button"
      w="64px"
      h="32px"
      bg={isDark ? "blue.700" : "blue.200"}
      borderRadius="18px"
      _dark={{ borderColor: "gray.50" }}
      onClick={onToggle}
      position="relative"
      aria-label={ariaLabel?.content}
    >
      {/* Visually hidden circle */}
      <Box
        borderRadius="50%"
        bg={isDark ? "blue.700" : "blue.200"}
        w="18px"
        h="18px"
        position="absolute"
        left="0"
        top="7px"
        zIndex="3"
      />
      {/* Main circle (sun/moon) */}
      <Box
        borderRadius="50%"
        bg="yellow.400"
        w="18px"
        h="18px"
        position="absolute"
        left="6px"
        top="7px"
        transform={`translateX(${isDark ? 0 : 34}px)`}
        transition="transform 0.5s"
        zIndex="2"
      />
      {/* Stars */}
      <Icon
        h="16px"
        w="16px"
        as={Stars}
        position="absolute"
        right="8px"
        top="7px"
        color="white"
        opacity={isDark ? 1 : 0}
        transition="opacity 0.3s"
        className={styles.stars}
        zIndex="1"
      />
      {/* Clouds */}
      <Icon
        h="16px"
        w="16px"
        position="absolute"
        left="8px"
        top="7px"
        color="white"
        as={Clouds}
        opacity={isDark ? 0 : 1}
        transition="opacity 0.4s"
        zIndex="4"
        className={styles.cloud}
      />
    </Box>
  );
};

export default ColorModeSwitcher;
