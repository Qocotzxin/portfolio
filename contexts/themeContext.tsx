import { hasWindow } from "@utils/window";
import {
  createContext,
  FC,
  PropsWithChildren,
  useContext,
  useEffect,
  useState,
} from "react";
import { ColorMode } from "types";

const COLOR_MODE_STORAGE_KEY = "color-mode";

enum CssVars {
  BodyLight = "var(--ce-color-body-light-bg)",
  BodyDark = "var(--ce-color-body-dark-bg)",
  BodyBg = "--ce-body-bg",
  BodyText = "--ce-body-text",
  InitialColorMode = "--initial-color-mode",
}

const BG_COLORS = {
  [ColorMode.light]: CssVars.BodyLight,
  [ColorMode.dark]: CssVars.BodyDark,
};

const TEXT_COLORS = {
  [ColorMode.light]: CssVars.BodyDark,
  [ColorMode.dark]: CssVars.BodyLight,
};

/**
 * Taken from https://www.joshwcomeau.com/react/dark-mode/
 */
export function getInitialColorMode(): ColorMode.light | ColorMode.dark {
  if (!hasWindow()) {
    return ColorMode.light;
  }

  const persistedColorPreference = window.localStorage.getItem(
    COLOR_MODE_STORAGE_KEY
  ) as ColorMode | null;
  const hasPersistedPreference = typeof persistedColorPreference === "string";

  if (hasPersistedPreference) {
    return persistedColorPreference;
  }

  const mql = window.matchMedia("(prefers-color-scheme: dark)");
  const hasMediaQueryPreference = typeof mql.matches === "boolean";
  if (hasMediaQueryPreference) {
    return mql.matches ? ColorMode.dark : ColorMode.light;
  }

  return ColorMode.light;
}

interface ThemeContextValue {
  colorMode: ColorMode;
  setColorMode: (value: ColorMode) => void;
  toggleColorMode: () => void;
}

const ThemeContext = createContext<ThemeContextValue | undefined>(undefined);

const ThemeProvider: FC<PropsWithChildren> = ({ children }) => {
  const [colorMode, rawSetColorMode] = useState<ColorMode>(
    getInitialColorMode()
  );

  const setColorMode = (value: ColorMode) => {
    rawSetColorMode(value);
    window.localStorage.setItem(COLOR_MODE_STORAGE_KEY, value);
  };

  const toggleColorMode = () => {
    const newColorMode =
      colorMode === ColorMode.dark ? ColorMode.light : ColorMode.dark;
    rawSetColorMode(newColorMode);
    window.localStorage.setItem(COLOR_MODE_STORAGE_KEY, newColorMode);
    const root = window.document.documentElement;
    root.style.setProperty(CssVars.BodyBg, BG_COLORS[newColorMode]);
    root.style.setProperty(CssVars.BodyText, TEXT_COLORS[newColorMode]);
  };

  useEffect(() => {
    const root = window.document.documentElement;
    const initialColorValue = root.style.getPropertyValue(
      CssVars.InitialColorMode
    );

    rawSetColorMode((initialColorValue as ColorMode) || ColorMode.light);
  }, []);

  return (
    <ThemeContext.Provider value={{ colorMode, setColorMode, toggleColorMode }}>
      {children}
    </ThemeContext.Provider>
  );
};

function useThemeContext() {
  const context = useContext(ThemeContext);

  if (context === undefined) {
    throw new Error("useThemeContext must be used within a ThemeProvider");
  }

  return context;
}

export { ThemeProvider, useThemeContext };
