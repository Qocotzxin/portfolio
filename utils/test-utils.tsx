import { ChakraProvider } from "@chakra-ui/react";
import { render, RenderOptions } from "@testing-library/react";
import { FC, PropsWithChildren, ReactElement } from "react";
import theme from "@theme";
import { HygraphProvider } from "contexts/hygraphContext";
import { Components, HygraphData } from "@models/hygraph";

export const hygraphMockData: Partial<HygraphData> = {
  skipLinks: [
    {
      text: "Skip to content",
    },
  ],
  ariaLabels: [
    {
      id: "1",
      metadata: "dark",
      component: Components.ColorModeSwitcher,
      content: "Switch to light mode",
    },
    {
      id: "2",
      metadata: "light",
      component: Components.ColorModeSwitcher,
      content: "Switch to dark mode",
    },
  ],
  navLinks: [
    {
      url: "test",
      order: 1,
      text: "Test",
    },
  ],
  languages: [
    { code: "en", displayName: "english", isActive: true },
    { code: "es", displayName: "spanish", isActive: true },
    { code: "nl", displayName: "dutch", isActive: false },
  ],
  footerLinks: [
    {
      url: "https://github.com",
      isActive: true,
      icon: "github",
      ariaLabel: "Visit Github",
    },
    {
      url: "https://test.com",
      isActive: false,
      icon: "github",
      ariaLabel: "Visit test",
    },
  ],
};

interface AllTheProvidersProps {
  hygraphData?: Partial<HygraphData>;
}

export const AllTheProviders: FC<PropsWithChildren & AllTheProvidersProps> = ({
  children,
  hygraphData = hygraphMockData,
}) => {
  return (
    <ChakraProvider theme={theme}>
      <HygraphProvider hygraphData={hygraphData}>{children}</HygraphProvider>
    </ChakraProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
  wrapperProps?: AllTheProvidersProps
) =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllTheProviders {...wrapperProps}>{children}</AllTheProviders>
    ),
    ...options,
  });

export * from "@testing-library/react";
export { customRender as render };
