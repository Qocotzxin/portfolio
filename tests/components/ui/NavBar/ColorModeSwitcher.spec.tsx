import { useThemeContext } from "@contexts/themeContext";
import { fireEvent, render, screen } from "@utils/test-utils";
import ColorModeSwitcher from "@ui/NavBar/ColorModeSwitcher";
import { ColorMode } from "types";
import { useAriaLabel } from "@hooks/useAriaLabel";

jest.mock("@hooks/useAriaLabel", () => ({
  useAriaLabel: jest.fn(),
}));

jest.mock("@contexts/themeContext", () => {
  const actual = jest.requireActual("@contexts/themeContext");
  return {
    ...actual,
    useThemeContext: jest.fn(() => ({
      colorMode: "light",
      toggleColorMode: jest.fn(),
      setColorMode: jest.fn(),
    })),
  };
});

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

const useColorModeMock = useThemeContext as jest.MockedFunction<
  typeof useThemeContext
>;

describe("ColorModeSwitcher", () => {
  it("Should match snapshot.", () => {
    const { container } = render(<ColorModeSwitcher />);

    expect(container).toMatchSnapshot();
  });

  it("Should call useAriaLabel hook with metadata 'dark' if colorMode is light.", () => {
    render(<ColorModeSwitcher />);

    expect(useAriaLabel).toHaveBeenCalledWith({
      component: "ColorModeSwitcher",
      metadata: "dark",
    });
  });

  it("Should call useAriaLabel hook with metadata 'light' if colorMode is dark.", () => {
    useColorModeMock.mockImplementation(() => ({
      colorMode: ColorMode.dark,
      toggleColorMode: jest.fn(),
      setColorMode: jest.fn(),
    }));

    render(<ColorModeSwitcher />);

    expect(useAriaLabel).toHaveBeenCalledWith({
      component: "ColorModeSwitcher",
      metadata: "light",
    });
  });

  it("Should call toggleColorMode when clicking the button.", () => {
    const toggleColorModeMock = jest.fn();
    useColorModeMock.mockImplementation(() => ({
      colorMode: ColorMode.light,
      toggleColorMode: toggleColorModeMock,
      setColorMode: jest.fn(),
    }));

    render(<ColorModeSwitcher />);

    fireEvent.click(screen.getByRole("button"));
    jest.runAllTimers();

    expect(toggleColorModeMock).toHaveBeenCalled();
  });
});
