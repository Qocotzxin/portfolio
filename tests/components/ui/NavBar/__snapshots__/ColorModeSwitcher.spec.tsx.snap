// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorModeSwitcher Should match snapshot. 1`] = `
.emotion-0 {
  width: 64px;
  height: 32px;
  background: var(--chakra-colors-blue-200);
  border-radius: 18px;
  position: relative;
}

.chakra-ui-dark .emotion-0:not([data-theme]),
[data-theme=dark] .emotion-0:not([data-theme]),
.emotion-0[data-theme=dark] {
  border-color: var(--chakra-colors-gray-50);
}

.emotion-1 {
  border-radius: 50%;
  background: var(--chakra-colors-blue-200);
  width: 18px;
  height: 18px;
  position: absolute;
  left: 0px;
  top: 7px;
  z-index: 3;
}

.emotion-2 {
  border-radius: 50%;
  background: var(--chakra-colors-yellow-400);
  width: 18px;
  height: 18px;
  position: absolute;
  left: 6px;
  top: 7px;
  -webkit-transform: translateX(34px);
  -moz-transform: translateX(34px);
  -ms-transform: translateX(34px);
  transform: translateX(34px);
  -webkit-transition: -webkit-transform 0.5s;
  transition: transform 0.5s;
  z-index: 2;
}

.emotion-3 {
  width: 24px;
  height: 24px;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  width: 16px;
  height: 16px;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: var(--chakra-colors-white);
  position: absolute;
  right: 8px;
  top: 7px;
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  z-index: 1;
}

.emotion-4 {
  width: 24px;
  height: 24px;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  width: 16px;
  height: 16px;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: var(--chakra-colors-white);
  position: absolute;
  left: 8px;
  top: 7px;
  opacity: 1;
  -webkit-transition: opacity 0.4s;
  transition: opacity 0.4s;
  z-index: 4;
}

<div>
  <button
    aria-label="Switch to light mode"
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <svg
      class="chakra-icon chakra-icon stars emotion-3"
      fill="currentColor"
      focusable="false"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7.657 6.247c.11-.33.576-.33.686 0l.645 1.937a2.89 2.89 0 0 0 1.829 1.828l1.936.645c.33.11.33.576 0 .686l-1.937.645a2.89 2.89 0 0 0-1.828 1.829l-.645 1.936a.361.361 0 0 1-.686 0l-.645-1.937a2.89 2.89 0 0 0-1.828-1.828l-1.937-.645a.361.361 0 0 1 0-.686l1.937-.645a2.89 2.89 0 0 0 1.828-1.828l.645-1.937zM3.794 1.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387A1.734 1.734 0 0 0 4.593 5.69l-.387 1.162a.217.217 0 0 1-.412 0L3.407 5.69A1.734 1.734 0 0 0 2.31 4.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387A1.734 1.734 0 0 0 3.407 2.31l.387-1.162zM10.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732L9.1 2.137a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L10.863.1z"
      />
    </svg>
    <svg
      class="chakra-icon chakra-icon cloud emotion-4"
      fill="currentColor"
      focusable="false"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.473 9a4.5 4.5 0 0 0-8.72-.99A3 3 0 0 0 3 14h8.5a2.5 2.5 0 1 0-.027-5z"
      />
      <path
        d="M14.544 9.772a3.506 3.506 0 0 0-2.225-1.676 5.502 5.502 0 0 0-6.337-4.002 4.002 4.002 0 0 1 7.392.91 2.5 2.5 0 0 1 1.17 4.769z"
      />
    </svg>
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
